src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -159,22 +164,22 @@ void LoaderConstraintTable::add_loader_constraint(Symbol* name, InstanceKlass* k
   LoaderConstraint* constraint = new LoaderConstraint(klass, class_loader1, class_loader2);
 
   // The klass may be null if it hasn't been loaded yet, for instance while checking
-  // an parameter name to a method call.  We impose this constraint that the
-  // class that is eventually loaded much match between these two loaders.
-  LoaderConstraintEntry* entry = _loader_constraint_table.get(name);
-  if (entry == nullptr) {
-    LoaderConstraintEntry entry(constraint);
+  // a parameter name to a method call.  We impose this constraint that the
+  // class that is eventually loaded must match between these two loaders.
+  bool created;
+  ClassNameEntry* entry = _loader_constraint_table.put_if_absent(name, &created);
```

Comment:
Is this "empty" entry visible before we get to call initialize on it below? Why do we not construct the entry properly before adding it? Do we really expect concurrent additions in this case?

Reply:
There is a lock around all accesses to this table.  This is a nicer pattern because we don't have to allocate, allocate again, copy, deallocate each entries with copy constructors.  We could do something fancy with the copy constructor and destructor (move semantics I think), this just seems easier to understand without tricky C++ code.